lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2645,109 +2779,47 @@ transactionExpirySlot safeTimeInterpreter maybeTTL =
     defaultTTL :: NominalDiffTime = 7200  -- that's 2 hours
 
 constructTxMeta
-    :: forall ctx s k.
-        ( HasDBLayer IO s k ctx
-        )
-    => ctx
+    :: DBLayer IO s k
     -> WalletId
     -> TransactionCtx
     -> [(TxIn, Coin)]
     -> [TxOut]
     -> ExceptT ErrSubmitTransaction IO TxMeta
-constructTxMeta ctx wid txCtx inps outs = db & \DBLayer{..} -> do
+constructTxMeta DBLayer{..} wid txCtx inps outs =
     mapExceptT atomically $ do
-        cp <- withExceptT ErrSubmitTransactionNoSuchWallet
+        checkpoint <- withExceptT ErrSubmitTransactionNoSuchWallet
               $ withNoSuchWallet wid
               $ readCheckpoint wid
-        liftIO $
-            mkTxMetaWithoutSel (currentTip cp) txCtx inps outs
-  where
-    db = ctx ^. dbLayer @IO @s @k
-
-mkTxMetaWithoutSel
-    :: BlockHeader
-    -> TransactionCtx
-    -> [(TxIn, Coin)]
-    -> [TxOut]
-    -> IO TxMeta
-mkTxMetaWithoutSel blockHeader txCtx inps outs =
-    let
-        amtOuts = F.fold $ map TxOut.coin outs
-
-        amtInps
-            = F.fold (map snd inps)
-            & case txWithdrawal txCtx of
-                w@WithdrawalSelf{} -> Coin.add (withdrawalToCoin w)
-                WithdrawalExternal{} -> Prelude.id
-                NoWithdrawal -> Prelude.id
-    in return TxMeta
-       { status = Pending
-       , direction = if amtInps > amtOuts then Outgoing else Incoming
-       , slotNo = blockHeader ^. #slotNo
-       , blockHeight = blockHeader ^. #blockHeight
-       , amount = Coin.distance amtInps amtOuts
-       , expiry = Just (snd $ txValidityInterval txCtx)
-       }
-
-ourCoin
-    :: IsOurs s Address
-    => TxOut
-    -> s
-    -> Maybe Coin
+        let latestBlockHeader = currentTip checkpoint
+        let amountOut = F.fold $ map TxOut.coin outs
+            amountIn = F.fold (map snd inps)
+                & case txWithdrawal txCtx of
+                    w@WithdrawalSelf{} -> Coin.add (withdrawalToCoin w)
+                    WithdrawalExternal{} -> Prelude.id
+                    NoWithdrawal -> Prelude.id
+        let validity = txValidityInterval txCtx
+        pure $ mkTxMeta latestBlockHeader validity amountIn amountOut
+
+ourCoin :: IsOurs s Address => TxOut -> s -> Maybe Coin
 ourCoin (TxOut addr tokens) wState =
-    case fst (isOurs addr wState) of
-        Just{}  -> Just (TokenBundle.getCoin tokens)
-        Nothing -> Nothing
+    fst (isOurs addr wState) $> TokenBundle.getCoin tokens
 
--- | Construct transaction metadata for a pending transaction from the block
--- header of the current tip and a list of input and output.
---
--- FIXME: There's a logic duplication regarding the calculation of the transaction
--- amount between right here, and the Primitive.Model (see prefilterBlocks).
```

Comment:
This is a good idea though - we may potentially need it 😉 https://github.com/input-output-hk/cardano-wallet/pull/3618/files#r1055560804

Reply:
What I don't like about this comment:

1. This also applies to other `FIXME`, `TODO` comments without a JIRA issue or author - its not clear who and when will do it. Most likely it will be kept until the heat death of the universe and nobody will dare to remove the comment.

2. The link to `prefilterBlocks` is broken: https://github.com/input-output-hk/cardano-wallet/search?q=prefilterBlocks